# Model Building

### Objective

- Prepare the processed data for machine learning.
- Separate input features (`X`) and target (`y`).
- Split the data into training and testing sets.
- Train multiple regression models.
- Compare their performance and select the best model.

## Import libraries

In [1]:
# Import required libraries
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

## Load the processed dataset

In [2]:
# Load the final processed dataset
df = pd.read_csv("../Data/03_Processed/mental_health_processed.csv")

# Display the shape and first few rows
print("Dataset Shape:", df.shape)
df.head()

Dataset Shape: (4998, 77)


,onehot__Gender_Female,onehot__Gender_Male,onehot__Country_Processed_Afghanistan,onehot__Country_Processed_Andorra,onehot__Country_Processed_Argentina,onehot__Country_Processed_Australia,onehot__Country_Processed_Bangladesh,onehot__Country_Processed_Brazil,onehot__Country_Processed_Canada,onehot__Country_Processed_China,...,onehot__Purpose_Of_Use_Networking,onehot__Purpose_Of_Use_News,ordinal__Stress_Level,scaler__Age,scaler__Avg_Daily_Usage_Hours,scaler__Daily_Unlocks,scaler__Study_Hours,scaler__Physical_Activity_Hours,scaler__Sleep_Hours_Per_Night,Mental_Health_Score
0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,1.0,0.0,1.0,0.102425,-0.652077,-0.873996,0.911362,0.672706,0.053499,6.8
1,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.254100,-2.103165,-2.297383,2.438856,0.972460,1.609042,7.6
2,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,...,0.0,0.0,1.0,0.678263,-0.289305,-0.127302,0.605863,0.073199,0.053499,7.0
3,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,3.0,-1.625088,1.161782,1.132745,-1.227130,-0.076678,-1.010819,5.3
4,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,1.0,0.0,3.0,1.829938,1.464092,1.529427,-1.227130,-0.975938,-1.338302,4.4


### Key Insights

- The processed dataset contains the features prepared during feature engineering.
- `Mental_Health_Score` is the target variable.
- The dataset is ready for machine learning model development.

## Separate Features and Target

- `X` contains the input features used by the model.
- `y` contains the value we want the model to predict.
- Here, `Mental_Health_Score` is the target.

In [3]:
# Separate input features and target varibales

X = df.drop(columns=["Mental_Health_Score"])
y = df["Mental_Health_Score"]

print("Feature Shape:", X.shape)
print("Target Shape:", y.shape)

Feature Shape: (4998, 76)
Target Shape: (4998,)


## Train-Test Split

- The dataset is divided into two parts:
  - **Training data** → used to teach the model.
  - **Testing data** → used to check how well the model performs on new data.
- **80%** of the data is used for training.
- **20%** of the data is used for testing.

In [4]:
# Split the data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20, # This Keep 20% of the data for testing
    random_state=42
)

print("Training Features:", X_train.shape)
print("Testing Features:", X_test.shape)
print("Training Target:", y_train.shape)
print("Testing Target:", y_test.shape)

Training Features: (3998, 76)
Testing Features: (1000, 76)
Training Target: (3998,)
Testing Target: (1000,)


### Key Insights

- 80% of the data is used for training the models.
- 20% is kept aside for testing.
- The test data is not used during model training.
- This helps us check how well the model works on unseen student data.

In [5]:
# Import regression models
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor

## Create Regression Models

- Four regression models are selected for comparison.
   1. Linear Regression
   2. Decision Tree
   3. Random Forest
   4. Gradiant Boosting
- Each model will learn the relationship between the student features and `Mental_Health_Score`.
- Their performance will be compared after training.

# 1. Linear Regression Model

- Linear Regression is a simple model 
- It tries to find the relationship between the input features and target variable.
- It is used as a baseline model for comparison with other algorithms

### Linear Regression Model - Creation

In [6]:
# Import Linear Regression
from sklearn.linear_model import LinearRegression

# Create the Linear Regression model
linear_model = LinearRegression()

# Train the Linear Regression model
linear_model.fit(X_train,y_train)# .fit() = teach the model using the training data


print("Linear Regression Model Is Trained Successfully!")

Linear Regression Model Is Trained Successfully!


### Linear Regression - Prediction

In [7]:
# Make predictions using the trained Linear Regression model
y_pred_linear = linear_model.predict(X_test)

# Display the first 10 predictions
y_pred_linear[:10]

array([4.6178524 , 5.86523818, 5.0687083 , 4.92145016, 7.92213459,
       5.89286965, 4.99230723, 6.64795327, 6.45216915, 5.16817349])

- The trained model is used to predict 'Mental_Health_Score' for the test data
- 'X_test' contains data that was not used during training
- The predicted scores are stored in 'y_pred_linear'

### Linear Regression - Model Evaluation

In [8]:
# Import evaluation metrics 
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [9]:
# Calculate evaluation metrics for Linear Regression

linear_mae = mean_absolute_error(y_test,  y_pred_linear)

linear_rmse = np.sqrt(mean_squared_error(y_test, y_pred_linear))

linear_r2 = r2_score(y_test,y_pred_linear)

In [10]:
# Store the evaluation metrics for Linear Regression

linear_results = pd.DataFrame({
    "Model": ["Linear Regression"],
    "MAE" :[linear_mae],
    "RMSE":[linear_rmse],
    "R2 Score": [linear_r2]
})

linear_results

,Model,MAE,RMSE,R2 Score
0,Linear Regression,0.475815,0.613455,0.789197


### Key Insights

- Linear Regression achieved an **MAE of 0.4758**.
- The **RMSE is 0.6135**.
- The **R² Score is 0.7892**, meaning the model explains a substantial portion of the variation in `Mental_Health_Score`.
- This result will be used as a baseline for comparison with the other regression models.

# 2. Decision Tree

- Decision Tree Regression predicts the target by making a series of decisions based on the input features
- It can capture non-linear relationships between features and 'Mental_Health_Score'
- The data is divided into smaller groups until the model reaches a prediction 

### Decision Tree model - Creation

In [11]:
# Import Decision Tree Regressor 
from sklearn.tree import DecisionTreeRegressor

# Create the Decision Tree model
decision_tree_model = DecisionTreeRegressor(random_state= 42)

# Train the Decision Tree Model
decision_tree_model.fit(X_train, y_train)

print("Decision Tree Trained Successfully!")

Decision Tree Trained Successfully!


### Decision Tree - Prediction

In [12]:
# Make prediciton using the trained Decision Tree model
y_pred_tree = decision_tree_model.predict(X_test)

# Display the first 10 predictions
y_pred_tree[:10]

array([5.3, 6.1, 3.8, 5.7, 7.3, 6. , 5.8, 7. , 7. , 5. ])

- The trained Decision Tree model is used to predict 'Menta_Health_Score'
- Predictions are made using the unseen test features('X_test')
- The predicted scores are stored in 'y_pred_tree'

### Decision Tree - Evaluation

In [13]:
# Calculate evaluation metrics for Decision Tree

tree_mae = mean_absolute_error(y_test,y_pred_tree)

tree_rmse = np.sqrt(mean_squared_error(y_test,y_pred_tree))

tree_r2 = r2_score(y_test, y_pred_tree)

In [14]:
# Store results in DataFrame
tree_results = pd.DataFrame({
    "Model": ["Decision Tree"],
    "MAE" :[tree_mae],
    "RMSE" :[tree_rmse],
    "R2 Score": [tree_r2]
})

tree_results

,Model,MAE,RMSE,R2 Score
0,Decision Tree,0.3228,0.51581,0.850964


### Key Insights

- Decision Tree achieved an **MAE of 0.3228**
- The **RMSE** is **0.5158**
- The **R2** Score is **0.8510**
- The final model will be selected after comparing all four models

# 3. Random Forest Regression

- Random Forest is made up of many Decision Trees 
- Each tree makes its own prediction
- The predictions from all trees are combined to produce the final prediction
- It can capture complex and non-linear relationships in the data

### Random Forest Regression model - Creation

In [15]:
# Import Random Forest Regressor 
from sklearn.ensemble import RandomForestRegressor

# Create the Random Forest model
random_forest_model = RandomForestRegressor(random_state=42)

# Train the Random Forest model
random_forest_model.fit(X_train, y_train)

print("Random Forest Model Trained Successfully!")



Random Forest Model Trained Successfully!


### Random Forest - Prediction

In [16]:
# Make predictions using the trained Random Forest model
y_pred_rf = random_forest_model.predict(X_test)

# Display the first 10 predictions
y_pred_rf[:10]

array([4.924, 5.994, 3.998, 5.577, 7.475, 5.986, 5.234, 6.892, 6.994,
       4.915])

- The trained Random Forest model is used to predict 'Mental_Health_Score'
- Predictions are made using the unseed features('X_test')
- The predicted scores are stored in 'y_pred_rf'

### Random Forest - Evaluation

In [17]:
# Calculate evaluation metrics for Random Forest

rf_mae = mean_absolute_error(y_test, y_pred_rf)

rf_rmse = np.sqrt(mean_squared_error(y_test,y_pred_rf))

rf_r2 = r2_score(y_test, y_pred_rf)

In [18]:
# Store results in a DataFrame
rf_results = pd.DataFrame({
    "Model" : ["Random Forest"],
    "MAE" : [rf_mae],
    "RMSE" : [rf_rmse],
    "R2 Score": [rf_r2]
})

rf_results

,Model,MAE,RMSE,R2 Score
0,Random Forest,0.277949,0.38565,0.91669


### Key Insights

- Random Forest achieved an **MAE of 0.277949**
- The **RMSE** is **0.38565**
- The **R2** is **0.91669**
- Random Forest currently performs better than Linear Regression and Decision Tree
- The final model will be selected after evaluating Gradient Boosting

# Gradient Boosting Regression 
- Gradient Boosting build multiple decision trees one after another 
- Each new tree tries to improve the errors made by previous trees 
- It can capture complex and non-linear relationships in the data 
- The final prediction combines the contributions of all the trees


### Gradient Boosting Regression Model - Creation 

In [21]:
# Import Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor

# Create Gradient Boost Model
gradient_boosting_model = GradientBoostingRegressor(
    random_state=42
)

# Train Gradient Boost Model
gradient_boosting_model.fit(X_train,y_train)

print("Gradient Boosting model trained successfully!")

Gradient Boosting model trained successfully!


### Gradient Boost - Prediction

In [26]:
# Make predictions using the Gradient Boosting model
y_pred_grad = gradient_boosting_model.predict(X_test)

# Display the first 10 predictions
y_pred_grad[:10]

array([4.92655473, 5.58678025, 4.7976616 , 4.9691428 , 8.03609155,
       6.14842545, 4.97398265, 6.68570092, 6.75895357, 5.0429672 ])

### Gradient Boost - Evaluation

In [28]:
# Calculate evaluation metrics for Gradient Boost

gb_mae = mean_absolute_error(y_test, y_pred_grad)

gb_rmse = np.sqrt(mean_squared_error(y_test,y_pred_grad))

gb_r2 = r2_score(y_test, y_pred_grad)

In [29]:
gb_results = pd.DataFrame({
    "Model" : ["Gradient Boost"],
    "MAE" : [gb_mae],
    "RMSE" : [gb_rmse],
    "R2 Score" : [gb_r2]
})

gb_results

,Model,MAE,RMSE,R2 Score
0,Gradient Boost,0.410558,0.529632,0.84287


### Key Insights

- Gradient Boosting achieved an **MAE of 0.410558**
- The **RMSE** is **0.529632**
- The **R2 Score** is **0.84287**
- Gradient Boosting performs better than Linear Regression but worse than Decision Tree and Random Forest on the test data
- Among the four models evaluated so far, **Random Forest shows the besst overall performance!**

## Final Model Comparison

In [42]:
# Combine the evaluation results of all four models
model_comparison = pd.concat([
    linear_results, tree_results, rf_results, gb_results
] , ignore_index = True)

# Display the final comparison
model_comparison.sort_values(ascending=False, by= "R2 Score").reset_index(drop=True)

# df.sort_values(ascending=False, by= "column_name") --> to sort values by specific column


# df.reset_index(drop=True) --> Discard the old, jumbled index and reset it to a clean 0, 1, 2... sequence

,Model,MAE,RMSE,R2 Score
0,Random Forest,0.277949,0.385650,0.916690
1,Decision Tree,0.322800,0.515810,0.850964
2,Gradient Boost,0.410558,0.529632,0.842870
3,Linear Regression,0.475815,0.613455,0.789197


### Key Insights

- Four regression models were evaluated on the same test dataset
- **Random Forest achieved the lowest MAE (0.2779)** and **lowest RMSE (0.3857)**
- Random Forest also achieved the **highest R² Score (0.9167)**
- Therefore, **Random Forest is the best-performing model among the four models tested**
- We will use Random Forest as the candidate model for the next stage of model development